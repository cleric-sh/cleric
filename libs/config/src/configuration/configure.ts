import {Selector, select} from './select';

export interface ILogger {
  (message: string): void;
}

export interface IConfigArgs {
  logger?: ILogger;
}

export function configure<TArgs extends IConfigArgs, TConfig extends Object>(
  config: (args: TArgs) => TConfig,
  defaultArgs: TArgs
) {
  return function <TSelector extends Selector<TConfig>>(
    requires: TSelector,
    args: TArgs = defaultArgs
  ) {
    const configuration = config(args);
    const {logger} = args;

    const [selected, loggable] = select(configuration, requires, !!logger);

    if (logger) {
      const json = JSON.stringify(loggable, null, 2);
      logger(`SELECTED CONFIGURATION: ${json}`);
    }

    return selected;
  };
}
