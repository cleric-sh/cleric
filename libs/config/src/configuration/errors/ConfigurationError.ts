export class ConfigurationError extends Error {
  constructor(public readonly message: string) {
    super(message);
  }
}
