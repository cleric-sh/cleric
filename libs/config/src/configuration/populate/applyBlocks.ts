import {Class} from 'Class/Class';
import {BLOCK_METADATA_KEY} from '../decorators/Block';
import {getMetadata} from '../utils/reflect/getMetadata';
import {SourcePaths} from './getSourcePaths';
import {removeSourcesFromCursor} from './removeSourcesFromCursor';

export function applyBlocks(target: Class | Object, propertyKey: string, current: SourcePaths) {
  const blockKeys = getMetadata(BLOCK_METADATA_KEY, target, propertyKey);
  if (blockKeys) {
    current = removeSourcesFromCursor(current, blockKeys);
  }
  return current;
}
