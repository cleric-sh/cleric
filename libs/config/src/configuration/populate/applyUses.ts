import {Misc} from 'ts-toolbelt';

import {SourcePaths, getSourcePaths} from './getSourcePaths';

export function applyUses(target: Misc.JSON.Object, propertyKey: string, current: SourcePaths) {
  const propertyKeys = getSourcePaths(target, propertyKey);

  const propertyValue = target[propertyKey];

  if (typeof propertyValue === 'object' && propertyValue !== null) {
    const propertyClassKeys = getSourcePaths(propertyValue);
    return {...current, ...propertyClassKeys, ...propertyKeys};
  }

  return {...current, ...propertyKeys};
}
