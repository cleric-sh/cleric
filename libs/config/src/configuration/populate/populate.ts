import {Class} from 'Class/Class';
import {Parameters} from 'Class/Parameters';
import {Misc} from 'ts-toolbelt';
import {Sources} from '../sources';
import {getSourcePaths} from './getSourcePaths';
import {populateObjectFromSources} from './populateObjectFromSources';

export function populate<TClass extends Class>(
  from: Sources,
  ctor: TClass,
  ...args: Parameters<TClass>
) {
  const settings = new ctor(...args) as Misc.JSON.Object;
  const sources = getSourcePaths(ctor);
  populateObjectFromSources(from, settings, sources, []);
  return settings as InstanceType<TClass>;
}
