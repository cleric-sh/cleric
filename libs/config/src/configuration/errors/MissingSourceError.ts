import {ConfigurationError} from './ConfigurationError';

export class MissingSourceError extends ConfigurationError {
  constructor(sourceKey: string, pathFromRoot: string[]) {
    const pathString = pathFromRoot.join('.');
    const message = `Configuration setting '${pathString}' expects source '${sourceKey}' to be provided, but it wasn't.`;
    super(message);
  }
}
