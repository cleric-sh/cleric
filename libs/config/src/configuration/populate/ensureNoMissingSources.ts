import {MissingSourceError} from '../errors/MissingSourceError';
import {SourceKey, Sources} from '../sources';
import {getSource} from './getSource';
import {SourcePaths} from './getSourcePaths';

export function ensureNoMissingSources(
  sources: Sources,
  cursor: SourcePaths,
  pathFromRoot: string[]
) {
  for (const key in cursor) {
    const _key = key as SourceKey;
    const source = getSource(sources, _key);
    if (!source) throw new MissingSourceError(key, pathFromRoot);
  }
}
