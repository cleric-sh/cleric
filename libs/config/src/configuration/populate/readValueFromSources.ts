import {Misc} from 'ts-toolbelt';
import {MissingSourceError} from '../errors/MissingSourceError';
import {Sources} from '../sources';
import {SourcePaths} from './getSourcePaths';

export function readValueFromSources(
  sources: Sources,
  target: Partial<Misc.JSON.Object>,
  sourcePaths: SourcePaths,
  pathFromRoot: string[]
) {
  // Step through each source in the order defined in 'sources'.
  for (const entry of sources) {
    const [key, source] = entry;
    const sourcePath = sourcePaths[key];

    // Skip this iteration if the cursor doesn't need to read from this source.
    if (!sourcePath) continue;

    if (!source) {
      throw new MissingSourceError(key, pathFromRoot);
    }

    const value = source.get(target, ...sourcePath);

    if (value) return value;
  }
  return undefined;
}
