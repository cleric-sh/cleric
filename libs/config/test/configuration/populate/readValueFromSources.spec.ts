import {MissingSourceError} from '../../../src/configuration/errors/MissingSourceError';
import {Sources} from '../../../src/configuration/populate';
import {readValueFromSources} from '../../../src/configuration/populate/readValueFromSources';

describe('readValueFromSources', () => {
  it('throws Error when property requires source but one is not provided', () => {
    const sources: Sources = [['ProcessEnv', undefined]];
    try {
      readValueFromSources(sources, {}, {ProcessEnv: []}, ['foo']);
    } catch (e) {
      expect(e).toBeInstanceOf(MissingSourceError);
      expect(e.message).toMatchSnapshot();
    }
  });
});
