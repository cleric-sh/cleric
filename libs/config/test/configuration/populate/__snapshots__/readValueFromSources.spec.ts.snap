// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readValueFromSources throws Error when property requires source but one is not provided 1`] = `"Configuration setting 'foo' expects source 'ProcessEnv' to be provided, but it wasn't."`;
